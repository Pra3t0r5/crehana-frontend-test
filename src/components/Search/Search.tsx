import { FC, useEffect, useState } from "react";
import { Box, IconButton, TextField } from "@mui/material";
import SearchIcon from "@mui/icons-material/Search";
import ClearIcon from "@mui/icons-material/Clear";
import { useCountriesQuery } from "../../generated/graphql";
import { DATASOURCE } from "../../constants";
import { CountryRowProps } from "../Country/Country";
import "./Search.css";
import { SearchProps } from "../../pages/CountriesPage/CountriesPage";

export interface SearchBarProps extends SearchProps {
  onLoading: (loading: boolean) => void;
  onComplete: (rows: CountryRowProps[]) => void;
}

function escapeRegExp(value: string): string {
  return value.replace(/[^\w_]/g, "");
}

const Search: FC<SearchBarProps> = ({ onSearchParamsChange, searchParams, onLoading, onComplete }) => {
  const [searched, setSearched] = useState<string>("");
  const { data, isFetching } = useCountriesQuery(
    DATASOURCE,
    searchParams,
    {
      queryKey: ["searchParams", JSON.stringify(searchParams)],
    }
  );

  const requestSearch = (searchedVal: string) => {
    const filter= {
      code: { regex: searchedVal },
    };
    onSearchParamsChange(filter);
    setSearched(searchedVal);
  };

  const handleTextInputChange = (event: any) => {
    const escaped = escapeRegExp(event.target.value);
    setSearched(escaped);
    requestSearch(escaped);
  };

  const cancelSearch = () => {
    setSearched("");
    requestSearch("");
  };

  useEffect(() => {
    if (data?.countries) {
      onComplete(data?.countries || []);
    }
  }, [data?.countries]);

  useEffect(() => {
    onLoading(isFetching);
  }, [isFetching]);

  return (
    <Box
      className="Search"
      data-testid="Search"
      sx={{
        pb: 0,
      }}
    >
      <TextField
        variant="standard"
        value={searched}
        onChange={handleTextInputChange}
        placeholder="Searchâ€¦"
        InputProps={{
          startAdornment: <SearchIcon fontSize="small" />,
          endAdornment: (
            <IconButton
              title="Clear"
              aria-label="Clear"
              size="small"
              style={{ visibility: searched ? "visible" : "hidden" }}
              onClick={cancelSearch}
            >
              <ClearIcon fontSize="small" />
            </IconButton>
          ),
        }}
        sx={{
          width: {
            xs: 1,
            sm: "auto",
          },
          m: (theme) => theme.spacing(1, 0.5, 1.5),
          "& .MuiSvgIcon-root": {
            mr: 0.5,
          },
          "& .MuiInput-underline:before": {
            borderBottom: 1,
            borderColor: "divider",
          },
        }}
      />
    </Box>
  );
};

export default Search;
