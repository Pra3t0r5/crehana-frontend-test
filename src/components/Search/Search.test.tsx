import React from "react";
import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import Search from "./Search";
import { QueryClient, QueryClientProvider } from "react-query";
import renderer from 'react-test-renderer';

describe("<Search />", () => {
  test("it should mount", () => {
    const queryClient = new QueryClient();
    const setRows = jest.fn();
    const setLoading = jest.fn();
    const set = jest.fn();
    const searchParams = {};
    render(
      <QueryClientProvider contextSharing={true} client={queryClient}>
        <Search
          onComplete={setRows}
          onLoading={setLoading}
          onSearchParamsChange={set}
          searchParams={searchParams}
        />
      </QueryClientProvider>
    );

    const search = screen.getByTestId("Search");

    expect(search).toBeInTheDocument();
  });

  test("it should match snapshot", () => {
    const queryClient = new QueryClient();
    const setRows = jest.fn();
    const setLoading = jest.fn();
    const set = jest.fn();
    const searchParams = {};
    const tree = renderer.create(
      <QueryClientProvider contextSharing={true} client={queryClient}>
        <Search
          onComplete={setRows}
          onLoading={setLoading}
          onSearchParamsChange={set}
          searchParams={searchParams}
        />
      </QueryClientProvider>
      ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
