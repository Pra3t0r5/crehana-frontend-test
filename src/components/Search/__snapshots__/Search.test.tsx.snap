// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> it should match snapshot 1`] = `
<div
  className="Search MuiBox-root css-8atqhb"
  data-testid="Search"
>
  <div
    className="MuiFormControl-root MuiTextField-root css-1pupl87-MuiFormControl-root-MuiTextField-root"
  >
    <div
      className="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd css-1ptx2yq-MuiInputBase-root-MuiInput-root"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
        data-testid="SearchIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInput-input MuiInputBase-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-1x51dt5-MuiInputBase-input-MuiInput-input"
        disabled={false}
        id=":r0:"
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Searchâ€¦"
        required={false}
        type="text"
        value=""
      />
      <button
        aria-label="Clear"
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "visibility": "hidden",
          }
        }
        tabIndex={0}
        title="Clear"
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
          data-testid="ClearIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
